android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java

```diff
@@ -90,79 +87,95 @@ private boolean evaluatePredicate(JsonElement user,
     }
 
     // TODO: 11.06.2021 need a unit test
-    private boolean evaluateRule(JsonElement user,
+    boolean evaluateRule(JsonElement user,
                                  JsonObject query,
                                  JsonObject rule,
                                  JsonObject passedRules,
                                  JsonObject failedRules) {
-        
+
         boolean result;
 
-        if(rule.has("combinator")){
+        if (rule.has("combinator")) {
             // No need to add groups to pass/failed rule sets here. Their children results will be merged up
             // via recursion.
             // eslint-disable-next-line no-use-before-define
             return evaluatePredicate(user, rule, passedRules, failedRules);
-        }else{
+        } else {
             result = evaluateFilter(user, rule);
         }
 
         // Any other rule is also a terminating branch in our recursion tree, so we add rule id to pass/fail rule set
-        if(result) {
-            passedRules.addProperty("id",query.get("id").getAsString());
-            passedRules.addProperty("index",rule.get("index").getAsString());
-            passedRules.addProperty("field",rule.get("field").getAsString());
-
-        }else{
-            failedRules.addProperty("id",query.get("id").getAsString());
-            failedRules.addProperty("index",rule.get("index").getAsString());
-            failedRules.addProperty("field",rule.get("field").getAsString());
+        if (result) {
+            passedRules.addProperty("id", query.get("id").getAsString());
+            if (rule.has("index"))
+                passedRules.addProperty("index", rule.get("index").getAsString());
+            passedRules.addProperty("field", rule.get("field").getAsString());
+
+        } else {
+            failedRules.addProperty("id", query.get("id").getAsString());
+            if (rule.has("index"))
+                failedRules.addProperty("index", rule.get("index").getAsString());
+            failedRules.addProperty("field", rule.get("field").getAsString());
         }
 
         return result;
     }
 
-    // TODO: 11.06.2021 need a unit test
-    private boolean evaluateFilter(JsonElement user, JsonObject rule) {
-        JsonElement value = valueFromKey(user, rule.get("field").getAsJsonObject());
+     boolean evaluateFilter(JsonElement user, JsonObject rule) {
+        JsonElement value = valueFromKey(user, rule.get("field").getAsString());
 
-        if (rule.get("operator").getAsString().startsWith("kv_") && value.isJsonNull() ) {
+        if (rule.get("operator").getAsString().startsWith("kv_") || value.isJsonNull()) {
```

Comment:
We do not use 'kv_' operators any more

Reply:
yes thanks, at that moment I did not know that they were not used 
i'll fix it

Reply:
fixed